### PR TITLE
[DONT MERGE] ThemeGenerator for Internal Code Review

### PR DESCRIPTION
**Pull request by wenyaoj2**
ThemeGenerator algorithm for variant theme

### CODE REVIEW

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+  AlphaColors,
+  type Greys
+} from '@fluentui/react-migration-v8-v9';
+import * as lodash from 'lodash';
+import { contrast, hex_to_sRGB } from './ThemeDesigner/src/colors';
+
+const standardContrastRatio: number = 4.5;
+const whiteHex: string = '#ffffff';
+const blackHex: string = '#000000';
+const greyReverse: Record<string, Greys> = Object.keys(grey).reduce((acc: any, key: any) => {
```
Comment (Leolewin) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:31]: Not use any

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.foregroundColor]],
+    darkMode ? whiteHex : '#323130',
+    darkMode,
+    false,
+    darkMode
+  );
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.backgroundColor]],
+    darkMode ? '#323130' : whiteHex,
```
Comment (Leolewin) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:48]: '#323130' can be set as global const variable

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+  type Greys
+} from '@fluentui/react-migration-v8-v9';
+import * as lodash from 'lodash';
+import { contrast, hex_to_sRGB } from './ThemeDesigner/src/colors';
+
+const standardContrastRatio: number = 4.5;
+const whiteHex: string = '#ffffff';
+const blackHex: string = '#000000';
+const greyReverse: Record<string, Greys> = Object.keys(grey).reduce((acc: any, key: any) => {
+  acc[grey[key]] = key;
```
Comment (RongqiZ) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:32]: lodash.invert?

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+      theme[token] = shift(theme[token], !isInverted, 4);
+    }
+  }
+
+  let color: IColor = getColorFromString(brand[80])!;
+  color = updateA(color, 5);
+  theme['colorNeutralBackground1'] = rgbaToHex(color, isInverted);
+}
+
+function getStrongVariant(theme: Theme, brand: BrandVariants, isInverted?: boolean) {
```
Comment (Leolewin) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:173]: `isInverted` is not used in this function

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+function getNeutralVariant(theme: Theme, isInverted: boolean) {
+  for (const token in theme) {
+    if (token.startsWith('colorNeutralBackground') && !isTokenNameContains(token, ['Alpha'])) {
+      const num: number | undefined = getNumberFromToken(token);
+      if (!num) {
+        continue;
+      }
+      if (num >= 4 && num <= 5) {
```
Comment (RongqiZ) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:103]: num === 4 || num === 5

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+    ...invertObject(grey, 'Grey'),
+    ...invertObject(brandVariants, 'Brand'),
+    ...invertObject(blackAlpha, 'BlackAlpha'),
+    ...invertObject(whiteAlpha, 'WhiteAlpha'),
+    whiteHex: 'white',
+    blackHex: 'black'
+  };
+}
+
+// convert colorBrandBackground = '#0067B7' TO colorBrandBackground = 'Brand80'
```
Comment (cool-sp) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:403]: this comment is not right?

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+
+const standardContrastRatio: number = 4.5;
+const whiteHex: string = '#ffffff';
+const blackHex: string = '#000000';
+const greyReverse: Record<string, Greys> = Object.keys(grey).reduce((acc: any, key: any) => {
+  acc[grey[key]] = key;
+  return acc;
+}, {});
+
+export const getV8Theme = (primaryColor: string, darkMode = false) => {
```
Comment (cool-sp) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:36]: why we need this function

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+  const themeRules = themeRulesStandardCreator();
+
+  ThemeGenerator.setSlot(
+    themeRules[BaseSlots[BaseSlots.foregroundColor]],
+    darkMode ? whiteHex : '#323130',
```
Comment (ymyqwe) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:41]: Use a const and name #323130

File: packages/variant-theme-v9/src/ThemeGenerator.ts (hunk 5 above)
Comment (cool-sp) [reply packages/variant-theme-v9/src/ThemeGenerator.ts:103]: +!

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+function getNeutralVariant(theme: Theme, isInverted: boolean) {
+  for (const token in theme) {
+    if (token.startsWith('colorNeutralBackground') && !isTokenNameContains(token, ['Alpha'])) {
+      const num: number | undefined = getNumberFromToken(token);
+      if (!num) {
+        continue;
+      }
+      if (num >= 4 && num <= 5) {
+        if (token.endsWith('hover')) {
+          theme[token] = shift(theme[token], !isInverted, 12);
```
Comment (cool-sp) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:105]: this reminds me that we should put some high level comments at the top to describe the general design principles.

File: packages/variant-theme-v9/src/ThemeGenerator.ts (hunk 5 above)
Comment (zeedomD) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:103]: magic number?

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+      }
+    } else if (
+      (token.startsWith('colorNeutralStroke') &&
+        !isTokenNameContains(token, ['Alpha', 'OnBrand', 'AccessibleSelected', 'InvertedDisabled'])) ||
+      (isTokenNameContains(token, ['Foreground']) &&
+        isTokenNameContains(token, ['Neutral']) &&
+        !isTokenNameContains(token, ['Brand'])) ||
+      isTokenNameContains(token, ['Subtle'])
+    ) {
+      theme[token] = shift(theme[token], !isInverted, 4);
```
Comment (cool-sp) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:120]: I think some of the code is from general pinciples and some of them is just a fix-up. can we figure out a way to distinguish them?

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+  const index: number = greyReverse[color];
+  return darken ? grey[Math.max(index - degree, 0)] : grey[Math.min(index + degree, 100)];
+}
+
+function getNumberFromToken(token: string): number | undefined {
+  const nums = token.match(/\d+/);
+  return nums ? parseInt(nums[0]) : undefined;
+}
+
+function rgbaToHex(color: IColor, isInverted: boolean): string {
```
Comment (ymyqwe) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:348]: Self invented or use something from color.js?

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+function getStrongVariant(theme: Theme, brand: BrandVariants, isInverted?: boolean) {
+  // Todo: dark mode support
+  setNeutralWithBrand(theme, brand);
+  const colorOnBrand: string = determineColorOnBrand(theme.colorNeutralBackground1);
+  const alphas: Record<AlphaColors, string> = colorOnBrand === black ? blackAlpha : whiteAlpha;
+
+  setColorsOnBrand(theme, brand, colorOnBrand, alphas);
+}
+
+function setNeutralWithBrand(theme: Theme, brand: BrandVariants) {
```
Comment (XinyangWang1024) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:182]: called in getStrongVarient, why name it setNeturalWithBrand

File: packages/variant-theme-v9/src/ThemeGenerator.ts [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+  const g: number = parseInt(color.substring(2, 4), 16); // hexToG
+  const b: number = parseInt(color.substring(4, 6), 16); // hexToB
+  const uicolors: number[] = [r / 255, g / 255, b / 255];
+  const c: number[] = uicolors.map((col) => {
+    if (col <= 0.03928) {
+      return col / 12.92;
+    }
+    return Math.pow((col + 0.055) / 1.055, 2.4);
+  });
+  const L: number = 0.2126 * c[0] + 0.7152 * c[1] + 0.0722 * c[2];
```
Comment (ymyqwe) [inline packages/variant-theme-v9/src/ThemeGenerator.ts:226]: Can you add some comments and params return value explanation to this function?

**Pull request closed without merging**